<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 3.2//EN">
<html>
<head>
    <meta name="Generator" CONTENT="Cosmo Create 1.0.3">
    <meta http-equiv="Content-Type" content="text/html; charset=iso-8859-1">
    <title>MultiSeq Bioinformatics Analysis Environment</title>
<link rel="stylesheet" type="text/css"
href="style.css" />
<style type="text/css">
<!--
.style1 {font-family: Arial, Helvetica, sans-serif}
-->
</style>
</head>
<body>
<table width="100%" border="0" cellspacing="0" cellpadding="0">
<tr>
<td valign="top" align="left">
<table width="100%" cellspacing="5">
<tr>
<td valign="top">
<h2 class="left">MultiSeq </h2>
<p class="left"><strong>MultiSeq</strong> is a unified bioinformatics analysis environment that allows one to organize, display, and analyze both sequence and structure data for proteins and nucleic acids. Special emphasis is placed on analyzing the data within the framework of evolutionary biology.</p>
<p class="left">In any publication of scientific results based completely or in part on the use of MultiSeq, please reference:</p>
<p class="left">
Elijah Roberts, John Eargle, Dan Wright, and Zaida Luthey-Schulten.<br>
MultiSeq: Unifying sequence and structure data for evolutionary analysis.<br>
<strong>BMC Bioinformatics, 2006, 7:382.</strong>
</p>

<h2 class="left"><A HREF="mseq_manual.pdf">User's Guide</A></h2>
<p class="left">
Visit the <a href="http://www.scs.illinois.edu/~schulten/multiseq/">Luthey-Schulten
group MultiSeq home page</a> for additional information and tutorials.
</p>
</td>
<td valign="top">
<img src="Overview.png" width="400" height="201"></td>
</tr>
</table>
</td>
</tr>
<tr><td><p>&nbsp;</p>
    <hr width="100%"></td>
</tr>
<tr>
<td valign="top" align="left">
<table width="100%" cellspacing="5">
<tr>
  <td width="50%" valign="top"><p><strong>Download<br>
  </strong>MultiSeq is included with <a href="http://www.ks.uiuc.edu/Research/vmd/">VMD</a> </p>
    <p>(Note: MultiSeq is located in the Extensions menu as Analysis-&gt;MultiSeq)</p>    </td>
  <td width="0" valign="top">&nbsp;</td>
  <td width="50%" valign="top"><p><strong>Learn</strong><br>
    A paper describing MultiSeq titled <a href="http://www.biomedcentral.com/1471-2105/7/382/abstract">"MultiSeq: Unifying sequence and structure data for evolutionary analysis."</a> has been released in BMC Bioinformatics.</p>
    <p>A tutorial entitled <a href="http://www.scs.illinois.edu/~schulten/tutorials/evolution/index.html">&quot;Evolution of Biomolecular Structure</a>&quot; is available for learning how to use MultiSeq.</p>
    </td>
</tr>
<tr>
  <td colspan="3" valign="top">
    <hr width="100%">    </td>
  </tr>
<tr>
  <td width="50%" valign="top"><p><strong>Features</strong></p>    </td>
  <td width="0" valign="top">&nbsp;</td>
  <td width="50%" valign="top">&nbsp;</td>
</tr>
<tr align="center" valign="bottom">
  <td width="50%">&nbsp;</td>
  <td width="0">&nbsp;</td>
  <td width="50%">&nbsp;</td>
</tr>
<tr align="center" valign="bottom">
  <td width="50%"><img src="SequenceStructure.png" width="375" height="153"></td>
  <td width="0">&nbsp;</td>
  <td width="50%"><img src="BlastResultViewer.png" width="375" height="151"></td>
</tr>
<tr>
  <td width="50%" align="left" valign="top"><p align="justify"><strong>Combine Sequence and Structure </strong>data to gain  insight into evolutionary changes in sequence, structure, and function. Compare sequence measures alongside structural measures with just a few mouse clicks. </p></td>
  <td width="0" align="left" valign="top">&nbsp;</td>
  <td width="50%" align="left" valign="top"><p align="justify"><strong>Quickly and Easily Import Data</strong> from many popular files formats or with an integrated BLAST search. Filter search results by sequence identity to eliminate duplicates or by taxonomy to focus a search. </p></td>
</tr>
<tr align="center" valign="bottom">
  <td width="50%">&nbsp;</td>
  <td width="0">&nbsp;</td>
  <td width="50%">&nbsp;</td>
</tr>
<tr align="center" valign="bottom">
  <td width="50%">&nbsp;</td>
  <td width="0">&nbsp;</td>
  <td width="50%">&nbsp;</td>
</tr>
<tr align="center" valign="bottom">
  <td width="50%"><img src="Infoviewer.png" width="357" height="215"></td>
  <td width="0">&nbsp;</td>
  <td width="50%"><img src="Align.png" width="375" height="165"></td>
</tr>
<tr>
  <td width="50%" valign="top"><p align="justify"><strong>Automatically Find Metadata </strong>for sequences and structures from published databases. Use this information (such as taxonomic classification or enzymatic function) during all steps of the analysis process. View and edit metadata via the electronic notebook. </p></td>
  <td width="0" valign="top">&nbsp;</td>
  <td width="50%" valign="top"><p align="justify"><strong>Align Sequences and
  Structures </strong>using popular alignment algorithms: MAFFT or ClustalW for sequences and STAMP for structures. Perform multiple alignments or profile alignments on entire molecules or just select regions. </p></td>
</tr>
<tr>
  <td width="50%" valign="middle">&nbsp;</td>
  <td width="0" valign="top">&nbsp;</td>
  <td width="50%" valign="top">&nbsp;</td>
</tr>
<tr>
  <td width="50%" valign="middle">&nbsp;</td>
  <td width="0" valign="top">&nbsp;</td>
  <td width="50%" valign="top">&nbsp;</td>
</tr>
<tr align="center" valign="bottom">
  <td width="50%"><img src="Grouping.png" width="374" height="149"></td>
  <td width="0">&nbsp;</td>
  <td width="50%"><img src="QR.png" width="375" height="150"></td>
</tr>
<tr>
  <td width="50%" valign="top"><p align="justify"><strong>Organize Data </strong>into custom groupings to ease the process of analyzing large amounts of data. Automatically group by taxonomy to help put the data into an evolutionary context. </p></td>
  <td width="0" valign="top">&nbsp;</td>
  <td width="50%" valign="top"><p align="justify"><strong>Eliminate Bias and Redundancy </strong>by running the QR algorithms for sequence and structure. </p></td>
</tr>
<tr align="center" valign="bottom">
  <td>&nbsp;</td>
  <td>&nbsp;</td>
  <td>&nbsp;</td>
</tr>
<tr align="center" valign="bottom">
  <td>&nbsp;</td>
  <td>&nbsp;</td>
  <td>&nbsp;</td>
</tr>
<tr align="center" valign="bottom">
  <td width="50%"><img src="Phylotree.png" width="229" height="200"></td>
  <td width="0">&nbsp;</td>
  <td width="50%"><img src="Plotter.png" width="206" height="200"></td>
</tr>
<tr>
  <td width="50%" valign="top"><p align="justify"><strong>View and Manipulate Phylogenetic Trees </strong> representing your data by either computing distance trees on the fly or loading precomputed trees from other phylogenetic packages. </p></td>
  <td width="0" valign="top">&nbsp;</td>
  <td width="50%" valign="top"><p align="justify"><strong>Plot Data Measures </strong>on a per residue basis with the built-in plotter. </p></td>
</tr>
<tr align="center" valign="bottom">
  <td>&nbsp;</td>
  <td>&nbsp;</td>
  <td>&nbsp;</td>
</tr>
<tr align="center" valign="bottom">
  <td width="50%">&nbsp;</td>
  <td width="0">&nbsp;</td>
  <td width="50%">&nbsp;</td>
</tr>
<tr>
  <td valign="top"><p><strong>Plus...</strong></p>    </td>
  <td valign="top">&nbsp;</td>
  <td valign="top">&nbsp;</td>
</tr>
<tr>
  <td width="50%" valign="top"><p align="justify"><strong>Work with Nucleic Acid </strong>sequences and structures  using many of the same techniques as for proteins. </p></td>
  <td width="0" valign="top">&nbsp;</td>
  <td width="50%" valign="top">&nbsp;</td>
</tr>
<tr>
  <td valign="top"><p align="justify"><strong>Export Data </strong>to many popular file formats or create  images of alignments, phylogenetic trees, or plots as publication quality PostScript graphics. </p></td>
  <td valign="top">&nbsp;</td>
  <td valign="top">&nbsp;</td>
</tr>
</table>
<hr width="100%">
</td>
</tr>
</table>
</td>
</tr>
</table>

<br>
</body>
</html>
